src/stirling/source_connectors/perf_profiler/perf_profiler_bpf_test.cc

```diff
@@ -251,6 +259,143 @@ class PerfProfileBPFTest : public ::testing::TestWithParam<std::filesystem::path
+
+        // try_emplace() looks up an existing entry in the strings table, or creates a new entry
+        // with the key provided. Here a new entry maps from symbol to next_location_id.
+        const auto [strings_iter, inserted] = strings.try_emplace(symbol, next_location_id);
+
+        if (inserted) {
+          // New symbol: create a new location, add it to the sample, and create a new symbol.
+
+          // For convenience & clarity, store the ids in locals, then compute their next values.
+          const uint64_t location_id = next_location_id;
+          const uint64_t string_id = next_string_id;
+          ++next_location_id;
+          ++next_string_id;
+
+          // Add a new "location" to the profile.
+          // Each sample is a sequence of locations. The locations, in their sequence, represent a
+          // stack trace. Each location may include an address and a reference to a mapping (useful
+          // if symbols are not included in the profile, enables post-hoc symbolization).
+          // Lacking both address and mapping here, we skip those.
+          auto location = profile.add_location();
+          location->set_id(location_id);
+
+          // To connect a location to a symbol, we need to go through a "line" and a "function".
+
+          // Add a "line" message to the location message.
+          // In our usage, this is essentially a pointer to a function message that points to a
+          // symbol. The function message may contain more information (e.g. starting line number).
+          auto line = location->add_line();
+          line->set_function_id(location_id);
+
+          // Add a "function" to the profile (to point to the string table).
+          auto function = profile.add_function();
+          function->set_id(location_id);
+
+          // Add a reference to the string table entry in the function message.
+          function->set_name(string_id);
+
+          // Add the string to the string table in the profile.
+          profile.add_string_table(std::string(symbol));
+
+          // Add the new location-id into the sample in the profile.
+          sample->add_location_id(location_id);
+
+          // Record the fact that we have written this string into the profile. Save the
+          // location-id because it is used if we find this symbol again.
+          strings[symbol] = location_id;
+        } else {
+          // Existing symbol.
+          // Just place the pre-existing location id into the sample.
+          const uint64_t location_id = strings_iter->second;
+          sample->add_location_id(location_id);
+        }
+      }
+    }
+
+    // Create a file.
+    std::ofstream ofs(FLAGS_pprof_pb_file, std::ios_base::out | std::ios_base::binary);
+
+    if (!ofs.is_open()) {
+      return error::ResourceUnavailable("Could not create or open file: $0.", FLAGS_pprof_pb_file);
+    }
+
+    // Serialize the pprof proto to that file.
+    if (!profile.SerializeToOstream(&ofs)) {
```

Comment:
`WritePProfProtoFile` seems to be called by multiple tests. My understanding is that the resulting pprof file ends up being whichever test ran last. Is that correct?

Reply:
Yes, that's true. In the example usage, I show the test case with a gtest filter (i.e. so that only one test is run and then the resulting pprof proto is unambiguous). I will figure out a way to either clarify the usage, or fail if the feature is abused, or uniquify the file name.

Reply:
What if we specified the file prefix and then suffixed test name or test index on the end of it? I think naming the flag that way will make the usage more obvious as well.